lib/semantics/check-do.cc

```diff
@@ -454,6 +394,7 @@ class DoContext {
     auto &doStmt{
         std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t)};
     currentStatementSourcePosition_ = doStmt.source;
+    context_.set_location(currentStatementSourcePosition_);
```

Comment:
`context` is not something owned by the `DoContext` visitor, if it is modified here, it should probably be restored afterwards to what it was so that the next visitor called by `SemanticsVisitor` does not have a bogus context.

In this case, it is most likely useless to do so, the location was just reset by `SemanticsVisitor` and there is most likely no visitor that will ever use the location before it is set again by `SemanticsVisitor`, but IMHO we should always restore upper level states that we modify and do not own just in case.

I would go for adding a method to `SemanticsContext` that uses `common::Restorer` and `common::ScopedSet` to locally modify `SemanticsContext.location_`.
See `parser::Message::SetLocation` for example:
https://github.com/flang-compiler/f18/blob/b5f48dfc461db37650a54aea3e29531b577fd871/lib/parser/message.h#L289
You may want second opinion on this proposal, maybe it is overkill and simply using `set_location` again at the end of your function is enough.

Reply:
Thanks for spotting this, Jean.  I didn't notice that calling ```set_location()``` would have these side effects.  The basic problem I'm trying to solve is to save the location of the ```DO CONCURRENT``` statement for use in error messages.  It seems to me that the most direct way to do that is just to put it into a data member of the ```DoChecker``` class.  Also, to avoid such side effects, the data member ```SemanticsContext &context_``` should be declared ```const```.

Let me know if you think there's a better way to go about this.

Reply:
You will probably not be able to declare it const because the `Say` methods are modifying the context.

Adding a member with the DO CONCURRENT location seems good to me. But if it is much more work than what you have now, I am not against modifying the context location as long as it is set back to what it was before.

Reply:
Thanks, Jean!